File: 0020-valid-parentheses/0020-valid-parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  stack = []
  open = {
    ')'=>'(',
    ']'=>'[',
    '}'=>'{',
  }
  s.each_char do |paren|
    if open.has_value?(paren)
      stack << paren
    elsif stack.pop != open[paren]
      return false 
    end
  end
  stack.length == 0
end

File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  const stack = [];
  const open = {
    ')': '(',
    ']': '[',
    '}': '{',
  };
  for (const p of s) {
    if ('([{'.includes(p)) {
      stack.push(p);
    } else if (stack.pop() !== open[p]) {
      return false;
    }
  }

  return !stack.length;
};